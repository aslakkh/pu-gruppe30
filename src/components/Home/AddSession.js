/**

 * Created by KasperKBerg on 28.03.2017.
 */

import React, { Component } from 'react';
import {Button, Modal, FormGroup, Label, Form, DropdownButton, MenuItem,ControlLabel} from 'react-bootstrap';
import { saveGoal } from '../../helpers/auth'
import {ref } from '../../config/constants'
import firebase from 'firebase';



let DatePicker = require("react-bootstrap-date-picker");
export default class AddSession extends Component{
    constructor(props){
        super(props);
        this.handleDateChange = this.handleDateChange.bind(this);
        this.handleGoalChange = this.handleGoalChange.bind(this);

        this.handleDropdownClick = this.handleDropdownClick.bind(this);
        this.validateState = this.validateState.bind(this);
        this.setHours = this.setHours.bind(this);
        this.setMinutes = this.setMinutes.bind(this);
        this.state = {
            courseID: this.props.courseID,
            date: new Date(new Date().getTime() + (24*60*60*1000)).toISOString(),
            show: false,
            hourSelected: 0,
            minSelected: 0,
            secondsPlanned: 0,
        }
    }

    handleDropdownClick(time) {
        this.setState({activeDropdownBtn: time})
    }

    setHours(num) {
        this.setState({
            hourSelected: num
        })
    }

    setMinutes(num) {
        this.setState({
            minSelected: num
        })
    }




    validateState(){
        //validates that date is after today
        //validates that goal is not empty
        let today = new Date(new Date);
        let stateDate = new Date(this.state.date);
        stateDate.setHours(0,0,0,0);
        return(stateDate.getTime() <= today.getTime() && this.state.goal.trim() !== "")
    }
    handleDateChange(e){

        this.setState({
            date: e,
        });
    }

    //handler for change in text area
    handleGoalChange(e){
        this.setState({
            goal: e.target.value,
        });
    }


    handleSave = (e) => {
        e.preventDefault(); //prevents default browser behaviour on click, whatever that means
        this.setState({secondsPlanned: (this.state.minSelected * 60) + (this.state.hourSelected * 3600)});
        //saveGoal(this.state.courseID, seconds);
        this.setState({
            show:false
        })

        this.castToFirebase();

    }


    castToFirebase(){
        const userUid = firebase.auth().currentUser.uid;
        const variabel = new Date(this.state.date)
        variabel.setMilliseconds(Math.random()*1000)
        const timeRef = ref.child('users/'+userUid+'/courses/'+this.state.courseID+'/sessions/'+ variabel.getTime());
        timeRef.set({time:(this.state.minSelected * 60) + (this.state.hourSelected * 3600),desc: "secondsPlanned"});
        this.setState({
            secondsPlanned: 0
        })
    }



    formValidateState(){

        if(this.state.goal.trim() !== ""){
            return 'success';
        }
        else{
            return 'error';
        }
    }

    datePickerValidateState(){
        let stateDate = new Date(this.state.date);
        stateDate.setHours(0,0,0,0);
        if(stateDate.getTime() <= Date.now()){
            return 'success';
        }
        else{
            return 'error';
        }
    }


    getView() {
        let items = [];
        items.push(<Label key={3}>Hours:</Label>);
        items.push(<DropdownButton key={4} title={this.state.hourSelected} id={'dropdown-basic'} >{this.getMenuItems('hours')}</DropdownButton>);
        items.push(<Label key={5}>Minutes:</Label>);
        items.push(<DropdownButton key={6} title={this.state.minSelected} id={'dropdown-basic'}>{this.getMenuItems('mins')}</DropdownButton>);

        return (
            <Form className="enter-form" inline>
                {items}
            </Form>
        )
    }


    getMenuItems(box) {
        let items = [];
        let max = 0;
        if (box==='hours') {
            for(let i=0; i<13; i++) {
                items.push(<MenuItem onClick={() => this.setHours(i)} key={i}>{i}</MenuItem>);
            }
        }
        if (box==='mins') {
            for(let i=0; i<60; i+=15) {
                items.push(<MenuItem onClick={() => this.setMinutes(i)} key={i}>{i}</MenuItem>);
            }
        }
        return items;
    }




    render() {
        let close = () => this.setState({show: false});

        return (
            <div className="main" style={{marginBottom: 10}}>
                <Button bsStyle="primary" bsSize="large" onClick={() => this.setState({show: true})}>
                    Set previous session
                </Button>
                <Modal className="modal" show={this.state.show} onHide={close} container={this}>
                    <Modal.Header closeButton>
                    </Modal.Header>
                    <Modal.Body>
                        <FormGroup
                            controlId="datePicker"

                            validationState={this.datePickerValidateState()}>
                            <ControlLabel>Date</ControlLabel>
                            <DatePicker id="example-datepicker" value={this.state.date} onChange={this.handleDateChange}/>
                        </FormGroup>
                    </Modal.Body>
                    <Modal.Body>
                        {this.getView()}

                    </Modal.Body>
                    <Modal.Footer>
                        <div>
                            <Button onClick={this.handleSave}>Save</Button>
                            <Button onClick={close}>Close</Button>
                        </div>
                    </Modal.Footer>
                </Modal>
            </div>
        );
    }
}