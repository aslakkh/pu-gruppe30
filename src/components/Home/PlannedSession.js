/**
 * Created by jan on 07/03/2017.
 */
import React, { Component } from 'react'
import {Button, ProgressBar, ListGroup, ListGroupItem, Grid,Col,Row} from 'react-bootstrap'
import firebase from 'firebase'


function Test(props){
    return(
        <ListGroup>
            <ListGroupItem>Link 1</ListGroupItem>
            <ListGroupItem onClick={ (link) => props.this.callbackParent(link)}>Link 2</ListGroupItem>
        </ListGroup>


    )
}




export default class PlannedSession extends Component{


    constructor(props){
        super(props)
        this.state={
            desc: "hei",
            emne: this.props.emne
        }

    }
componentWillMount(){
    let that = this
    const userUid = firebase.auth().currentUser.uid;
    firebase.database().ref().child('users/'+userUid+'/courses/'+that.state.emne+'/planned-sessions/').orderByValue().startAt().on('value', snap => {
        console.log(snap.val())
        that.setState({
            sessions: snap.val()
        })
        console.log(snap.val())
})
}

    handleClick(key){
        this.props.callbackParent(key);
    }



    render(){
        return(
            <ListGroup >
                {Object.keys(this.state.sessions).map((key) => {
                    return <ListGroupItem key={key} className="CoursesList">
                        {this.state.sessions[key].goal}
                        {console.log(key)}
                        {new Date(parseInt(key)).toISOString()}
                    </ListGroupItem>
                })}
            </ListGroup>
        )
    }
}