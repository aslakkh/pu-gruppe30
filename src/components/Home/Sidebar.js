import React, { Component } from 'react'
import {Tab,NavItem, Nav, Col,Row} from 'react-bootstrap'
import SpecificCourse from './specificCourse'

/*Maps the courses to make the sidebar and the site for the specific courses
*
*
* Changes:
* Change the way SpecificCourse is called, so you only load the course you are viewing.
*
* */




export default class Courses extends Component {
constructor(props) {
  super(props);
  this.state={
			
			courses: this.props.courses
		}
  
};

componentWillReceiveProps(nextProps){
  this.setState({
    courses: nextProps.courses
  });
}
    handleSelect(key){
        this.setState({
            key:key

        })

    }



componentWillMount(){
}
	render (){

    if(this.state.courses === undefined){ //checks if props are loaded
      return(
        <h4>ingenting</h4>
      );
    }
    else{
      return (

            <Tab.Container id="left-tabs" defaultActiveKey={(Object.keys(this.state.courses)[0])} activeKey={this.state.key}>
                <Row className="clearfix">
                <Col sm={3}>
                <Nav bsStyle="pills" className="navbar navbar-inverse" stacked>
                    {Object.keys(this.state.courses).map((emne,i) =>

                    <NavItem eventKey={emne} className="navbar-link" key={i} onSelect={this.setState({key:emne})} >
                        {emne}
                    </NavItem>


                    )}
                </Nav>
                </Col>
                <Col sm={9}>
                <Tab.Content animation>
                    {Object.keys(this.state.courses).map((emne,i) =>
                    <Tab.Pane eventKey={emne} key={i}>
                        <h1>{emne}</h1>
                        <SpecificCourse courseID={emne} course={this.state.courses[emne]}/>
                    </Tab.Pane>

                    )}
                </Tab.Content>
                </Col>
                </Row>
                </Tab.Container>









			)
    }
		
	}




}