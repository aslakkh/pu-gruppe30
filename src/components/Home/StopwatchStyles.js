import React from 'react';

export const styles = {
    timer: {
        textAlign: 'center',
        fontSize: '500%'
},
    greyText: {
        color:'#999999'
    },
    plannedText: {
        fontSize:'130%'
    },
    center: {
        textAlign: 'center',
        marginBottom: '10px',
    },
    margin: {
        marginTop: '4%'
    }

};



/*body {
    font-family: 'HelveticaNeue-UltraLight', 'Helvetica Neue UltraLight',
    'Helvetica Neueu', Helvetica, Arial, sans-serif;
    margin: 0;
    background-color: #efeff5;
}

.Stopwatch { text-align: center }
.Stopwatch-timer {
    font-size: 200px;
    font-weight: 100;
    line-height: 160px;

    margin: 0;

    border: solid #cecfd0;
    border-width: 1px 0;
    background-color: white;
}

.Stopwatch-laps {
    padding: 0 0 0 20px;
    list-style-type: none;
}

.Stopwatch-lap {
    font-size: 20px;
    line-height: 60px;
    padding-right: 20px;
    text-align: right;
    border-bottom: 1px solid #d9dae0;
}

.Stopwatch-lap strong {
    float: left;
    color: #7f8083;
    font-weight: 100;
}

.Stopwatch-lap strong::before { content: 'Lap ' }

.Stopwatch-btn {
    font-family: inherit;
    font-size: 20px;

    display: inline-block;

    width: 72px;
    height: 72px;
    margin: 24px;
    padding: 0;

    cursor: pointer;
    letter-spacing: 1px;

    border: 0;
    border-radius: 50%;
    outline: none;
    background: white;
}
.Stopwatch-app-timer {
    font-size: 200px;
}*/

