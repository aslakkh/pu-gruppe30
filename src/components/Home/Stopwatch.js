import React, { Component } from 'react';
import Progress from 'react-progressbar';
import {ref } from '../../config/constants'
import firebase from 'firebase';
import {Button, ProgressBar, ListGroup, Grid,Col,Row} from 'react-bootstrap'
import PlannedSession from './PlannedSession'
import AddSession from './AddSession'

const formattedSeconds = ((sec) => //formats to hh:mm:ss
Math.floor (sec/3600)+ ':' + Math.floor(sec / 60) + '.' + ('0' + sec % 60).slice(-2));


class Stopwatch extends Component {
    constructor(props) {
        super(props);

        this.state = {
            secondsElapsed: 0,
            lastClearedIncrementer: null,
            goal: 400,
            desc: ""
        };
        this.incrementer = null;
        this.started = false
    }

    castToFirebase(){
        console.log("desc");
        console.log(this.state.desc.value);
    const userUid = firebase.auth().currentUser.uid;
    const timeRef = ref.child('users/'+userUid+'/courses/'+this.state.emne+'/sessions/'+ Date.now());
    timeRef.set({time:this.state.secondsElapsed,desc: this.state.desc.value});
    this.setState({
        secondsElapsed: 0
    })
}
    handleStartClick() {
        if(!this.started){ //Makes sure the button isnt clicked twice
            this.started = true;
            this.incrementer = setInterval( () =>
                this.setState({
                    secondsElapsed: this.state.secondsElapsed + 1
                })
            , 1000);}
    }

    handleStopClick() {
        this.started = false;
        clearInterval(this.incrementer);
        this.setState({
            lastClearedIncrementer: this.incrementer
        });
    }




    componentWillMount(){
        this.setState ({
            secondsElapsed: 0,
            lastClearedIncrementer: null,
            goal: 400,
            emne: this.props.emne,
            theme: undefined
        })
        this.incrementer = null;
        this.started = false
        console.log(this.props.emne)

    }
    handleResetClick() {
        this.castToFirebase();
        clearInterval(this.incrementer);
    }
    onChildChanged(newState){
        console.log(newState)
        this.setState({theme:newState})

    }

    render() {
        return (
            <Grid fluid={true}>
                <Row>
                    <Col md={7}>
            <div className="stopwatch" >
                <h4>Planned Session: {this.state.theme}</h4>
                <h1 className="app-timer">{formattedSeconds(this.state.secondsElapsed)}</h1>

                {(this.state.secondsElapsed === 0 ||
                    this.incrementer === this.state.lastClearedIncrementer
                        ? <Button block className="btn btn-lg" onClick={this.handleStartClick.bind(this)}>start</Button>
                        : <Button block className="btn btn-lg" onClick={this.handleStopClick.bind(this)}>stop</Button>
                )}

                {(this.state.secondsElapsed !== 0 &&
                    this.incrementer === this.state.lastClearedIncrementer
                        ? <div>  <form><label>What did you do?{this.state.desc}</label>
                            <input className="form-control" ref={(desc) => this.state.desc = desc} placeholder="Description"/>
                                </form>
                            <Button block className="btn btn-lg" onClick={this.handleResetClick.bind(this)}>save</Button></div>
                        : null
                )}

            </div><div>
                        <AddSession courseID = {this.props.emne}/>
                    </div>

                    </Col>
                    <Col md={1}/>
            <Col md={4}>

                <div>
                    <PlannedSession emne={this.props.emne} callbackParent={(newState) => this.onChildChanged(newState) }/>
                </div>
</Col>
                </Row>
            </Grid>
        );
    }
}


export default Stopwatch;

