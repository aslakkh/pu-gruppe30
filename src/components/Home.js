import React, { Component } from 'react'
import Sidebar from './protected/Sidebar'
export default class Home extends Component {
  render () {
    return (
      <div>
      <Sidebar/>
        Home. Not Protected. Anyone can see this.
      </div>
    )
  } 
}