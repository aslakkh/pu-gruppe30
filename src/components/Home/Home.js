import React, { Component } from 'react'
import Sidebar from './Sidebar'
export default class Home extends Component {
  constructor(props){
    super(props);
    console.log("Home got: ");
    console.log(this.props.courses);
    this.state=({
        courses: this.props.courses,
        authed: this.props.authed,
        admin: this.props.admin
    })
  };
  componentWillReceiveProps(nextProps){
      console.log("Home Receives: ");
      console.log(nextProps.courses);
      this.setState({
          courses: nextProps.courses,
          authed: nextProps.authed,
          admin: nextProps.admin
      });
  }

  render () {
      if(this.state.authed && (this.state.admin != undefined || this.state.admin != null)){
          if(this.state.courses != undefined && this.state.courses != null){
            return(
              <Sidebar courses={this.state.courses}/>
          );
        }
        else{
            return <h4>No courses</h4>
        }
      }
    else{return (
      <div>
        Home. Not Protected. Anyone can see this.
      </div>
    )}
  } 
}