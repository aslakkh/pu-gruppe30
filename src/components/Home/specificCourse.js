import React, {Component} from 'react';
import Stopwatch from './Stopwatch'
import TabView from './TabView'

export default class SpecificCourse extends Component {
	constructor(props){
		super(props);
		this.state = ({
			courseID:props.courseID,
			points: undefined,
			course: props.course

		})


	}






	render(){
		return(

		<div>
			<TabView course={this.state.course} emne={this.state.courseID} />
		</div>)}






}