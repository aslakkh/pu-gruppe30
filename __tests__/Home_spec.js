/**
 * Created by jan on 19/04/2017.
 */
import React from 'react';
import renderer from 'react-test-renderer';
import Home from '../src/components/Home/Home.js';
import TestUtils from "react-addons-test-utils"

const courses={
  TES1000:{active: true,
      goals: {dailyGoal:{active: false,
          timeSet: 0,
          timeSpent: 0,
          value: 0},
          monthlyGoal:
              {active:false,
          timeSet:0,
          timeSpent:0,
          value:0},
          weeklyGoal:
              {active:false,
          timeSet:0,
          timeSpent:0,
          value:0}

  }
}};
const coursesNull=null;

describe("Home", function () {
   let renderer2;
   let renderer3;
   let renderer4;
   let component2;
   let component3;
   let component4;
   renderer2 = TestUtils.createRenderer();
   renderer2.render(
       <Home/>
   );
    component2 = renderer2.getRenderOutput();
   it("Not logged in", () => {
       expect(component2).toMatchSnapshot();
   });


    renderer3 = TestUtils.createRenderer();
    renderer2.render(
        <Home courses={courses} authed={true} admin={false}/>
    );
    component3 = renderer3.getRenderOutput();
    it("log in, with courses", () => {
        expect(component3).toMatchSnapshot();
    })
    renderer4 = TestUtils.createRenderer();
    renderer4.render(
        <Home courses={coursesNull} authed={true} admin={false}/>
    );
    component4 = renderer4.getRenderOutput();
    it("log in, no courses", () => {
        expect(component4).toMatchSnapshot();
    })
});
/*
test('renders home if not logged in', () => {
    const component = renderer.create(
        <Home />
    ).toJSON();
    expect(component).toMatchSnapshot(); //renders h4 if no courses in props
    const component2 = renderer.create(
        <Home courses={courses} authed={true} admin={true}/>
    ).toJSON();
    expect(component2).toMatchSnapshot();
})*/