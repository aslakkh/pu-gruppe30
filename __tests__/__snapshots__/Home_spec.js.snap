// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Not logged in 1`] = `
<div>
  Home. Not Protected. Anyone can see this.
</div>
`;

exports[`Home log in, no courses 1`] = `
<h4>
  No courses
</h4>
`;

exports[`Home log in, with courses 1`] = `null`;
